// advanced-command-handler/dist/classes/commands/ImageCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImageCommand = void 0;
const discord_js_1 = require("discord.js");
const discord_js_better_embed_1 = require("discord.js-better-embed");
const _1 = require("./");
/**
 * @see {@link https://ayfri.gitbook.io/advanced-command-handler/concepts/commands/templates}
 */
class ImageCommand extends _1.Command {
    /**
     * Send a local image in an embed from your files.
     *
     * @param options - The options.
     * @returns - The message with the embed sent.
     */
    async sendImageEmbed(options) {
        const channel = 'ctx' in options ? options.ctx.channel : options.channel;
        const link = options.link ?? options.path ?? '';
        const embed = discord_js_better_embed_1.BetterEmbed.fromTemplate('image', {
            ...options,
            image: link,
        });
        const attachment = new discord_js_1.MessageAttachment(link);
        if (options.path)
            embed.setImageFromFile(attachment);
        return await channel.send({ embeds: [embed] });
    }
    /**
     * Send a local image from your files.
     *
     * @param options - The options.
     * @returns - The message with the image sent.
     */
    async sendLocalImage(options) {
        const channel = 'ctx' in options ? options.ctx.channel : options.channel;
        return await channel.send({
            content: options.content,
            files: [options.path],
        });
    }
}
exports.ImageCommand = ImageCommand;
