// advanced-command-handler/dist/classes/AdvancedClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdvancedClient = void 0;
const discord_js_1 = require("discord.js");
const utils_1 = require("../utils");
class AdvancedClient extends discord_js_1.Client {
    /**
     * @param token - The token for the client.
     * @param options - The client options.
     * @see {@link https://discord.js.org/#/docs/main/stable/typedef/ClientOptions}
     */
    constructor(token, options) {
        super(options);
        this.token = token;
        utils_1.Logger.comment('Client initialized.', 'loading');
    }
    /**
     * Get the ID of the ClientUser.
     *
     * @returns - Returns the ID the ClientUser.
     */
    get id() {
        return this.user?.id;
    }
    /**
     * Get the mention of the ClientUser.
     */
    get mention() {
        return `<@${this.id}>`;
    }
    /**
     * Returns the ping of the websocket.
     */
    get ping() {
        return this.ws.ping;
    }
    /**
     * Returns the tag of the ClientUser.
     */
    get tag() {
        return this.user?.tag;
    }
    /**
     * Returns the username of the ClientUser.
     */
    get username() {
        return this.user?.username;
    }
}
exports.AdvancedClient = AdvancedClient;
