// argumentorum/dist/Cursor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cursor = void 0;
class Cursor {
    constructor(input) {
        this.input = input;
        this.index = -1;
    }
    get hasNext() {
        return this.index < this.input.length - 1;
    }
    get hasPrevious() {
        return this.index > 0;
    }
    consumeRemaining() {
        let result = '';
        while (this.hasNext)
            result += this.next();
        return result;
    }
    consumeWhile(predicate) {
        let result = undefined;
        while (this.hasNext && predicate(this.peekNext()))
            result = (result ?? '') + this.next();
        return result;
    }
    skipWhitespace() {
        if (this.peekNext() !== ' ')
            return false;
        while (this.peekNext() === ' ')
            this.next();
        return true;
    }
    next() {
        if (this.hasNext) {
            this.index++;
            return this.input[this.index];
        }
    }
    previous() {
        if (this.hasPrevious) {
            this.index--;
            return this.input[this.index];
        }
    }
    peek() {
        return this.input[this.index];
    }
    peekNext() {
        return this.hasNext ? this.input[this.index + 1] : undefined;
    }
    peekPrevious() {
        return this.hasPrevious ? this.input[this.index - 1] : undefined;
    }
}
exports.Cursor = Cursor;
