// advanced-command-handler/dist/utils/codeError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.codeError = void 0;
const discord_js_better_embed_1 = require("discord.js-better-embed");
const __1 = require("../");
/**
 * A function to use when a code error occurs in a command for example.
 *
 * @remarks
 * If the message author is an owner, it sends the error stack. <br>Do not use for sending string as errors.
 * @param ctx - The command context where the error come from.
 * @param error - The native error.
 * @returns - The error message sent.
 */
function codeError(ctx, error) {
    const embed = discord_js_better_embed_1.BetterEmbed.fromTemplate('complete', {
        client: ctx.client,
        color: 0xee2200,
        title: 'Code error :',
        description: (0, __1.cutIfTooLong)(error.stack ?? error.toString(), 2048),
    });
    __1.Logger.error(error, 'CodeError');
    if ((0, __1.isOwner)(ctx.user.id))
        return ctx.reply({ embed });
    return ctx.reply(`An error occurred while executing the \`${ctx.command.name}\` command.`);
}
exports.codeError = codeError;
