// advanced-command-handler/dist/utils/permissionUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isPermission = exports.permissionsError = void 0;
const discord_js_1 = require("discord.js");
const discord_js_better_embed_1 = require("discord.js-better-embed");
/**
 * A function to use when a user, or the client hasn't all the permissions needed.
 *
 * @param ctx - The command context where the permission where missing.
 * @param missingPermissions - The error.
 * @param fromClient - If the error is from the client.
 * @returns - The error message sent.
 */
function permissionsError(ctx, missingPermissions, fromClient = false) {
    const embed = discord_js_better_embed_1.BetterEmbed.fromTemplate('complete', {
        client: ctx.client,
        color: 0xee2200,
        title: 'Permissions error :',
        description: `${fromClient ? 'The bot is' : 'You are'} missing permissions :\n\`${missingPermissions.sort().join('`\n`')}\``,
    });
    if (ctx.command.usage)
        embed.addField('Syntax :', ctx.command.usage);
    return ctx.reply({ embed });
}
exports.permissionsError = permissionsError;
/**
 * Check if some permission is a valid permission that exists.
 *
 * @param permission - The permission to test.
 * @returns - Is the permission is a valid permission.
 */
function isPermission(permission) {
    return Object.keys(discord_js_1.Permissions.FLAGS).includes(permission);
}
exports.isPermission = isPermission;
